bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java

```diff
@@ -125,54 +126,51 @@ public void initialize() {
         }
 
         connector = new Ipx800DeviceConnector(configuration.hostname, configuration.portNumber, getThing().getUID());
-        parser = new M2MMessageParser(connector, this);
+        parser = Optional.of(new M2MMessageParser(connector, this));
 
         updateStatus(ThingStatus.UNKNOWN);
 
-        refreshJob = scheduler.scheduleWithFixedDelay(statusFile::read, 3000, configuration.pullInterval,
-                TimeUnit.MILLISECONDS);
+        refreshJob = Optional.of(scheduler.scheduleWithFixedDelay(statusFile::read, 3000, configuration.pullInterval,
+                TimeUnit.MILLISECONDS));
 
         connector.start();
     }
 
     @Override
     public void dispose() {
-        if (refreshJob != null) {
-            refreshJob.cancel(true);
-            refreshJob = null;
-        }
+        refreshJob.ifPresent(job -> job.cancel(true));
+        refreshJob = Optional.empty();
 
         if (connector != null) {
             connector.destroyAndExit();
         }
-        parser = null;
+        parser = Optional.empty();
 
         portDatas.values().stream().forEach(portData -> {
-            portData.destroy();
+            if (portData != null) {
+                portData.dispose();
+            }
         });
         super.dispose();
     }
 
     protected void discoverAttributes() {
-        final Map<String, String> properties = new HashMap<>();
+        updateProperties(Map.of(Thing.PROPERTY_VENDOR, "GCE Electronics", Thing.PROPERTY_FIRMWARE_VERSION,
+                statusFile.getElement(StatusEntry.VERSION), Thing.PROPERTY_MAC_ADDRESS,
+                statusFile.getElement(StatusEntry.CONFIG_MAC)));
 
-        properties.put(Thing.PROPERTY_VENDOR, "GCE Electronics");
-        properties.put(Thing.PROPERTY_FIRMWARE_VERSION, statusFile.getElement(StatusEntry.VERSION));
-        properties.put(Thing.PROPERTY_MAC_ADDRESS, statusFile.getElement(StatusEntry.CONFIG_MAC));
-        updateProperties(properties);
-
-        ThingBuilder thingBuilder = editThing();
         List<Channel> channels = new ArrayList<>(getThing().getChannels());
-
-        PortDefinition.asStream().forEach(portDefinition -> {
-            int nbElements = statusFile.getMaxNumberofNodeType(portDefinition);
-            for (int i = 0; i < nbElements; i++) {
-                createChannels(portDefinition, i, channels);
-            }
-        });
-
-        thingBuilder.withChannels(channels);
-        updateThing(thingBuilder.build());
+        if (channels.isEmpty()) {
```

Comment:
This is IMHO the wrong approach. You should create each dynamic channel but only after checking it does not yet exist.

Reply:
You're correct, I've enhanced the approach.